source/concurrent/SynchronizedQueue.ooc

```diff
@@ -48,34 +48,33 @@ SynchronizedQueue: class <T> extends Queue<T> {
 }
 
 BlockedQueue: class <T> extends SynchronizedQueue<T> {
-	_populated := WaitCondition new()
+	_waitLock : WaitLock
 	_canceled := false
-	init: func { super() }
+	init: func {
+		super()
+		this _waitLock = WaitLock new(_mutex)
+	}
 	free: override func {
-		this _populated free()
+		this _waitLock free()
 		super()
 	}
 	enqueue: override func (item: T) {
 		super(item)
-		this _populated signal()
+		this _waitLock wake()
 	}
 	cancel: func {
-		this _mutex lock()
-		this _canceled = true
-		this _mutex unlock()
-		this _populated broadcast()
+		this _waitLock with(|| this _canceled = true)
+		this _waitLock wakeAll()
 	}
 	wait: func (isOk := null as Bool*) -> T {
 		result: T = null
-		this _mutex lock()
-		while (this empty && !this _canceled)
-			this _populated wait(this _mutex)
+		this _waitLock lockWhen(|| !this empty || this _canceled)
```

Comment:
So, I would suggest something like
```ooc
populatedOrCanceled := func -> Bool { !this empty || this _canceled }
this _waitLock lockWhen(populatedOrCanceled)
```
to avoid the double `||`.